### DISCUSSION TITLE
Upgrading to latest pyright causes new error - false positive? (1.1.251 from 1.1.246)

### DISCUSSION THREAD

**Discussion by jab** [microsoft/pyright#3530] [Q&A]
As part of my monthly maintenance work, I just upgraded from pyright 1.1.246 to 1.1.251, the current latest stable release. The new version emits a new error at a couple call sites, unless I include the `__NO_KW: Mapping[str, Any]` workaround you can see in the first line of [the "upgrade pyright" commit](https://github.com/jab/bidict/commit/f31a6f88f3c9899808286767d97529f5175cad7c) (this is only a 5-line diff and it's very simple, so don't be scared to click through:).

Here are examples of the new error that is emitted without the workaround:

```console
  /Users/jab/src/bidict/bidict/_bidict.py:93:9 - error: Argument of type "dict[str, VT@BidictBase]" cannot be assigned to parameter "kw" of type "Mapping[str, VT@MutableBidict]" in function "_update"
    TypeVar "_VT_co@Mapping" is covariant
      Type "VT@BidictBase" cannot be assigned to type "VT@MutableBidict" (reportGeneralTypeIssues)
  /Users/jab/src/bidict/bidict/_bidict.py:179:13 - error: Argument of type "dict[str, VT@BidictBase]" cannot be assigned to parameter "kw" of type "Mapping[str, VT@MutableBidict]" in function "_update"
    TypeVar "_VT_co@Mapping" is covariant
      Type "VT@BidictBase" cannot be assigned to type "VT@MutableBidict" (reportGeneralTypeIssues)
```

Since pyright didn't emit any error before, and since the latest stable release of mypy also is not emitting any error without the workaround, I'm wondering if this is a regression in pyright?

If not, then:
1. Is there a better solution than the workaround I came up with?
2. Is there a bug in the mypy tracker for its associated false negative?

Thanks,
Josh

**Comment by erictraut** [comment]
If you think there's a bug, please file a bug report with a minimal self-contained code sample that exhibits the problem.

**Reply by jab** [reply]
I don't know enough about the recent changes to Pyright to have a good idea of whether this is due to a regression in Pyright. Before I invested more time in this, I wanted to do my best with the limited time I have to report something useful back to the Pyright team, which is much more familiar with the recent changes than I am, and can hopefully say with much more accuracy whether they suspect this is a regression or not based on glancing at the linked commit and excerpted output?

**Reply by jab** [reply]
If you already did take a glance, and do suspect a regression, then I'll be happy to file a bug report. I'm just not sure I'll have the time to include a minimized repro; I suspect that could take me 10x longer than it would take you, but I'll certainly give it a shot if you tell me you still need me to do that. And any hints on that would be much appreciated. (I'm guessing this has something to do with MutableMapping.update taking VT_co, but I've tried going down the VT_co route with bidict before and hit up against a lot more issues, possibly due to bugs in Pyright and mypy that were present at the time.)

**Reply by erictraut** [reply]
I took a quick look at it, and it appears that it might be a regression. I'll need to dig into it further when I have time. I've created an [issue](https://github.com/microsoft/pyright/issues/3534) to track it so it doesn't get lost.

**Reply by jab** [reply]
Thank you!

**Reply by jab** [reply]
Upgraded to 1.1.252, which includes the [fix for this issue](https://github.com/microsoft/pyright/commit/0dff59a90b410f090d90ec4c01edba84c64c0de2), and confirmed I can now remove the workaround from my code. Thanks for getting the fix out so quickly! And thanks for looking before I could come up with a minimized test case; looking at the one you came up with, I'm not sure I ever would have come up with that from what I was seeing.

Say, the bidict codebase seems like an unusually good corpus for catching bugs in type checkers, especially relative to its small size. E.g. pyright runs on the entire codebase in under 2 seconds. Would it make sense to consider adding a `pyright bidict` call to pyright's own CI, to help catch new bugs before they make it into release? (Or at least try it to see how well this works, and back it out if it turns out to not be worth it?) I'd be happy to put some time into helping set this up if you think it could save us time overall in the long run.

Thanks again for the quick fix and for the great work on pyright.

**Reply by hmc-cs-mdrissi** [reply]
A similar idea has been mentioned before in typeshed to introduce a pyright primer that tests pyright on some libraries to ensure typeshed changes don’t cause accidental regressions.  https://github.com/python/typeshed/issues/7642

Maybe bidict could be first library part of pyright primer.

**Reply by jab** [reply]
Interesting, thanks for the info, @hmc-cs-mdrissi. That would be an honor if bidict could help in this way.

Do you have any thoughts about this idea, @erictraut?

**Reply by erictraut** [reply]
@jab, are you interested in writing a "pyright-primer" — an analog to "mypy-primer" — that would test new combinations of pyright and typeshed stubs on existing code bases? I'd welcome that if you're volunteering.

**Reply by jab** [reply]
I was envisioning adding something not much more complicated than the following to pyright's CI:
```console
cd regression_tests/bidict
git clone http://github.com/jab/bidict latest
pyright ./latest/bidict > ./pyright-candidate.out
diff ./pyright-candidate.out ./pyright-expected.out  # exits nonzero if there's a diff
```
...and only some `pyright-expected.out` fixture would need to be checked into the pyright regression_tests dir. Any other projects that have also been good for discovering regressions in pyright besides bidict could be added similarly. But again, I imagined this first as something potentially low-effort/high-impact to try first, and only to be continued if it delivered on that potential.

If mypy and/or typeshed are already doing something like this (just not with bidict), maybe I should be starting there rather than with pyright? I started with pyright only because that's where bidict found a regression most recently, and also pyright's pace of new releases being so much faster could mean that's where I'd continue to hit regressions more often.

That said, I'm expecting my first child any day now (which is partially why it took me so long to write back here), so my chance to contribute to this idea anytime soon is rapidly diminishing. Sorry about that, and hope this discussion has at least been somewhat helpful.

**Reply by erictraut** [reply]
@jab, you may be interested to know that we now have "mypy_primer" hooked up in the pyright CI workflow! The author of mypy_primer (@hauntsaninja) graciously extended the tool to support pyright. I've configured pyright's CI scripts to run regression tests on a small number of open-source libraries including `bidict`. I plan to expand the list of libraries over time.

**Reply by jab** [reply]
Thanks for doing this, @erictraut and @hauntsaninja, and thanks for letting me know! Hope that having bidict included helps prevent future regressions. Always love to hear about different projects working together like this, and hope to follow along and contribute as much as I can.
